include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h

```diff
@@ -320,35 +361,61 @@ struct __parallel_merge_submitter_large<_IdType, _CustomName,
                         __start = __base_diagonals_sp_global_ptr[__diagonal_idx];
                     }
 
-                    __serial_merge(__rng1, __rng2, __rng3, __start.first, __start.second, __i_elem,
-                                   __nd_range_params.chunk, __n1, __n2, __comp);
+                    const auto __ends = __serial_merge(__rng1, __rng2, __rng3, __start.first, __start.second, __i_elem,
+                                                       __nd_range_params.chunk, __n1, __n2, __comp, __n);
+                    if constexpr (_OutSizeLimit{})
+                        if (__global_idx == __nd_range_params.steps - 1)
+                        {
+                            auto __res_ptr = _Storage::__get_usm_or_buffer_accessor_ptr(__result_acc);
+                            *__res_ptr = __ends;
+                        }
                 });
         });
     }
 
+    template <typename _Storage>
+    static constexpr auto
+    __get_acc(const _Storage& __base_diagonals_sp_global_storage, sycl::handler& __cgh)
+    {
+        if constexpr (_OutSizeLimit{})
+            return __base_diagonals_sp_global_storage.template __get_result_acc<sycl::access_mode::write>(
+                __cgh, __dpl_sycl::__no_init{});
+        else
+            return int{0};
+    }
+
   public:
     template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3, typename _Compare>
     auto
     operator()(_ExecutionPolicy&& __exec, _Range1&& __rng1, _Range2&& __rng2, _Range3&& __rng3, _Compare __comp) const
     {
-        assert(__rng1.size() > 0 || __rng2.size() > 0);
+        const _IdType __n1 = __rng1.size();
+        const _IdType __n2 = __rng2.size();
+        assert(__n1 > 0 || __n2 > 0);
+
+        const _IdType __n = std::min<_IdType>(__n1 + __n2, __rng3.size());
 
         _PRINT_INFO_IN_DEBUG_MODE(__exec);
 
         // Calculate nd-range parameters
-        const nd_range_params __nd_range_params = eval_nd_range_params(__exec, __rng1, __rng2);
+        const nd_range_params __nd_range_params = eval_nd_range_params(__exec, __rng1, __rng2, __n);
 
         // Create storage to save split-points on each base diagonal + 1 (for the right base diagonal in the last work-group)
-        auto __p_base_diagonals_sp_global_storage =
-            new __result_and_scratch_storage<_ExecutionPolicy, _split_point_t<_IdType>>(
+        using __val_t = _split_point_t<_IdType>;
+        __result_and_scratch_storage<_ExecutionPolicy, __val_t>* __p_base_diagonals_sp_global_storage = NULL;
+        if constexpr (_OutSizeLimit{})
+            __p_base_diagonals_sp_global_storage = new __result_and_scratch_storage<_ExecutionPolicy, __val_t>(
+                __exec, 1, __nd_range_params.base_diag_count + 1);
+        else
+            __p_base_diagonals_sp_global_storage = new __result_and_scratch_storage<_ExecutionPolicy, __val_t>(
```

Comment:
```suggestion
        auto __p_base_diagonals_sp_global_storage = new __result_and_scratch_storage<_ExecutionPolicy, __val_t>(
            __exec, _OutSizeLimit{} ? 1 : 0, __nd_range_params.base_diag_count + 1);

```

Reply:
I'm not sure that constexpr `_OutSizeLimit{}` is correct in the following line:
`_exec, _OutSizeLimit{} ? 1 : 0, __nd_range_params.base_diag_count + 1);`

Reply:
No problems, just checking it in run-time